acvm-repo/acvm/tests/solver.rs

```diff
@@ -722,3 +726,223 @@ fn memory_operations() {
 
     assert_eq!(witness_map[&Witness(8)], FieldElement::from(6u128));
 }
+
+// Solve the given BlackBoxFuncCall with witnesses: 1, 2 as x, y, resp.
+#[cfg(test)]
+fn solve_blackbox_func_call(
+    blackbox_func_call: impl Fn(
+        Option<FieldElement>,
+        Option<FieldElement>,
+    ) -> BlackBoxFuncCall<FieldElement>,
+    x: (FieldElement, bool), // if false, use a Witness
+    y: (FieldElement, bool), // if false, use a Witness
+) -> FieldElement {
+    let (x, x_constant) = x;
+    let (y, y_constant) = y;
+
+    let initial_witness = WitnessMap::from(BTreeMap::from_iter([(Witness(1), x), (Witness(2), y)]));
+
+    let mut lhs = None;
+    if x_constant {
+        lhs = Some(x);
+    }
+
+    let mut rhs = None;
+    if y_constant {
+        rhs = Some(y);
+    }
+
+    let op = Opcode::BlackBoxFuncCall(blackbox_func_call(lhs, rhs));
+    let opcodes = vec![op];
+    let unconstrained_functions = vec![];
+    let mut acvm =
+        ACVM::new(&StubbedBlackBoxSolver, &opcodes, initial_witness, &unconstrained_functions, &[]);
+    let solver_status = acvm.solve();
+    assert_eq!(solver_status, ACVMStatus::Solved);
+    let witness_map = acvm.finalize();
+
+    witness_map[&Witness(3)]
+}
+
+fn function_input_from_option(
+    witness: Witness,
+    opt_constant: Option<FieldElement>,
+) -> FunctionInput<FieldElement> {
+    opt_constant
+        .map(|constant| FunctionInput::constant(constant, FieldElement::max_num_bits()))
+        .unwrap_or(FunctionInput::witness(witness, FieldElement::max_num_bits()))
+}
+
+fn and_op(x: Option<FieldElement>, y: Option<FieldElement>) -> BlackBoxFuncCall<FieldElement> {
+    let lhs = function_input_from_option(Witness(1), x);
+    let rhs = function_input_from_option(Witness(2), y);
+    BlackBoxFuncCall::AND { lhs, rhs, output: Witness(3) }
+}
+
+fn xor_op(x: Option<FieldElement>, y: Option<FieldElement>) -> BlackBoxFuncCall<FieldElement> {
+    let lhs = function_input_from_option(Witness(1), x);
+    let rhs = function_input_from_option(Witness(2), y);
+    BlackBoxFuncCall::XOR { lhs, rhs, output: Witness(3) }
+}
+
+fn prop_assert_commutative(
+    op: impl Fn(Option<FieldElement>, Option<FieldElement>) -> BlackBoxFuncCall<FieldElement>,
+    x: (FieldElement, bool),
+    y: (FieldElement, bool),
+) -> (FieldElement, FieldElement) {
+    (solve_blackbox_func_call(&op, x, y), solve_blackbox_func_call(&op, y, x))
+}
+
+fn prop_assert_associative(
+    op: impl Fn(Option<FieldElement>, Option<FieldElement>) -> BlackBoxFuncCall<FieldElement>,
+    x: (FieldElement, bool),
+    y: (FieldElement, bool),
+) -> (FieldElement, FieldElement) {
+    (solve_blackbox_func_call(&op, x, y), solve_blackbox_func_call(&op, y, x))
```

Comment:
this is not associativity, this is identical to  commutativity

Reply:
@guipublic Thanks: I updated the test and it's now failing. Could you take a look before I open an issue for it?

Reply:
I think this `x.1 ^ y.1` (and also the others) should rather be a AND: `x.1 & y.1`
The result is constant only if both inputs are constants.

Reply:
@guipublic That `^` only applies to the `use_constant: bool` that determines whether to use a constant or witness, so it shouldn't affect the result. If you like, I can extend the test to accept fresh random `use_constant: bool`'s instead of using `^` of the `use_constant` of the inputs.

Reply:
I just updated the test and it's still failing, with all `use_constant=false`, i.e. all `FunctionInput::witness`'s:
```bash
Test failed: assertion failed: `(left == right)` 
  left: `1283670960019916003771447463403780437091510857904449510892463880104724571074`,
 right: `944443117489026359168770680972523757203580128233348138908887269072586843074` at acvm-repo/acvm/tests/solver.rs:871.
minimal failing input: x = (
    7250061818685696343216493786193839024302863820576605192710784820741603895966,
    false,
), y = (
    -3405112893702401459380202402317574365406674719909456358229164704216370206212,
    false,
), z = (
    2⁴×300433304925226584082054090820988662109061133690344416189530074644868631050,
    false,
), use_constant_xy = false, use_constant_yz = false
	successes: 0
	local rejects: 0
	global rejects: 0
```

Reply:
It's not clear to me what would be the issue. I think we can debug directly on the issue created from the failing tests, that would be easier.